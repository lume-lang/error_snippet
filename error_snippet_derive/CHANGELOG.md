# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.5](https://github.com/lume-lang/error_snippet/compare/error_snippet_derive-v0.1.4...error_snippet_derive-v0.1.5) - 2025-06-14

### Added

- *(derive)* allow merging spans and labels into single field

## [0.1.4](https://github.com/lume-lang/error_snippet/compare/error_snippet_derive-v0.1.3...error_snippet_derive-v0.1.4) - 2025-06-12

### Fixed

- *(derive)* correctly handle lifetimes on error structs

## [0.1.3](https://github.com/lume-lang/error_snippet/compare/error_snippet_derive-v0.1.2...error_snippet_derive-v0.1.3) - 2025-05-27

### Added

- add `#[cause]` attribute

## [0.1.2](https://github.com/lume-lang/error_snippet/compare/error_snippet_derive-v0.1.1...error_snippet_derive-v0.1.2) - 2025-05-11

### Fixed

- fix!(derive): use dynamic errors for #[related]

## [0.1.1](https://github.com/lume-lang/error_snippet/compare/error_snippet_derive-v0.1.0...error_snippet_derive-v0.1.1) - 2025-05-11

### Other

- move repository to `lume-lang`
