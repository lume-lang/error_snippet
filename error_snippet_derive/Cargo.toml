[package]
name = "error_snippet_derive"
version = "0.1.6"
authors = ["Max T. Kristiansen <me@maxtrier.dk>"]
description = "Derive macros for `error_snippet`."
repository = "https://github.com/lume-lang/error_snippet"
license = "MIT"
edition = "2018"
rust-version = "1.85.0"

[lib]
proc-macro = true

[dependencies]
proc-macro2 = "1.0.83"
quote = "1.0.35"
syn = "2.0.87"

[dev-dependencies]
insta = "=1.43.1"
