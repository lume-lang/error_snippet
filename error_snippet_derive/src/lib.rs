use diagnostic::Diagnostic;
use quote::quote;
use syn::{parse_macro_input, DeriveInput};

mod args;
mod diagnostic;
mod fields;
mod fmt;
mod tokens;

#[proc_macro_derive(Diagnostic, attributes(diagnostic, span, label, related, cause))]
pub fn derive_diagnostic(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let input = parse_macro_input!(input as DeriveInput);
    let cmd = match Diagnostic::from(input) {
        Ok(cmd) => cmd.tokens().unwrap_or_else(|err| err.to_compile_error()),
        Err(err) => return err.to_compile_error().into(),
    };

    quote!(#cmd).into()
}
