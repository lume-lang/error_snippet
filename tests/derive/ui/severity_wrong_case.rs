use error_snippet_derive::Diagnostic;

#[derive(Debug, Diagnostic)]
#[diagnostic(message = "some help", severity = help)]
struct Foo {}

fn main() {}
