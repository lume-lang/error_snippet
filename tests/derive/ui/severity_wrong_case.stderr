error[E0599]: no variant or associated item named `help` found for enum `Severity` in the current scope
 --> tests/derive/ui/severity_wrong_case.rs:4:48
  |
4 | #[diagnostic(message = "some help", severity = help)]
  |                                                ^^^^ variant or associated item not found in `Severity`
  |
help: there is a variant with a similar name
  |
4 - #[diagnostic(message = "some help", severity = help)]
4 + #[diagnostic(message = "some help", severity = Help)]
  |
